pkg/gce-pd-csi-driver/node.go

```diff
@@ -365,7 +367,38 @@ func (ns *GCENodeServer) NodeGetVolumeStats(ctx context.Context, req *csi.NodeGe
 }
 
 func (ns *GCENodeServer) NodeExpandVolume(ctx context.Context, req *csi.NodeExpandVolumeRequest) (*csi.NodeExpandVolumeResponse, error) {
-	return nil, status.Error(codes.Unimplemented, fmt.Sprintf("NodeExpandVolume is not yet implemented"))
+	volumeID := req.GetVolumeId()
+	if len(volumeID) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "ControllerExpandVolume volume ID must be provided")
+	}
+	capacityRange := req.GetCapacityRange()
+	reqBytes, err := getRequestCapacity(capacityRange)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("ControllerExpandVolume capacity range is invalid: %v", err))
+	}
+
+	volKey, err := common.VolumeIDToKey(volumeID)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("ControllerExpandVolume volume ID is invalid: %v", err))
+	}
+
+	devicePath, err := ns.getDevicePath(volumeID, "")
+	if err != nil {
+		return nil, status.Error(codes.Internal, fmt.Sprintf("Error when getting device path for %s: %v", volumeID, err))
+	}
+
+	resizer := resizefs.NewResizeFs(ns.Mounter)
+	resized, err := resizer.Resize(devicePath, "")
```

Comment:
what happens if it's already resized?

Reply:
This will not work for file systems that require device mount path though.. I think you will have to pass deviceMountPath to the `Resize` function as well. it can be extracted from node expand request.

Reply:
*cough* *cough* - xfs

Reply:
also semi related: https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/issues/305 and https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/issues/306

Reply:
> what happens if it's already resized?

I believe `resize2fs` just returns success and therefore this `resizer.Resize` will return success. Which is pretty much what we expect.
Not sure about the behavior of the `xfs` tool but I assume its the same/similar since it also doesn't take `size` as an arg, it just resizes to the size of the underlying partition